### DISCUSSION TITLE
自动同步豆瓣图书封面到 Notion 的读书笔记

### DISCUSSION THREAD

**Discussion by LeetaoGoooo** [LeetaoGoooo/discussion-blog#14] [随笔]
**TL;DR**

可以通过以下脚本实现同步封面到 Notion 读书笔记

```python
from typing import List, TypedDict
import requests
import os

database_id = os.getenv("database_id") # 读书笔记对应到数据库id
notion_version = os.getenv("notion_version", "2022-02-22") # 固定参数
token = os.getenv("token") # 自己的integrations的token

if not all([database_id, token]):
    raise ValueError("neither token or database id can't be none")

notion_headers = {
    'Authorization': f'Bearer {token}',
    'Content-Type': 'application/json',
    'Notion-Version': notion_version
}


class EmptyCoverPage(TypedDict):
    page_id: str
    book_name: str


def _get_image_from_douban(book_name: str) -> str:
    resp = requests.get(
        f'https://book.douban.com/j/subject_suggest?q={book_name}', headers={"User-Agent": "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; AcooBrowser; .NET CLR 1.1.4322; .NET CLR 2.0.50727)"})
    if resp.status_code != 200 or not resp.json():
        return
    resp_json = resp.json()
    return resp_json[0]["pic"]


def _get_pages_with_empty_cover(database_id: str) -> List[EmptyCoverPage]:
    url = f"https://api.notion.com/v1/databases/{database_id}/query"
    payload = {
        "filter": {
            "property": "封面图", # 封面列的名称
            "files": {
                "is_empty": True
            }
        }
    }
    resp = requests.post(url, json=payload, headers=notion_headers)
    if resp.status_code != 200:
        return []
    resp_json = resp.json()
    return [EmptyCoverPage(page_id=item["id"], book_name=item["properties"]["书名"]["title"][0]["text"]["content"]) for item in resp_json["results"]]


def _set_cover(cover: str, page_id: str):
    url = f"https://api.notion.com/v1/pages/{page_id}"
    payload = {
        "properties": {
            "封面图": {
                "files": [{
                    "name": cover,
                    "type": "external",
                    "external": {
                        "url": cover
                    }
                }]
            }
        }
    }
    resp = requests.patch(url, json=payload, headers=notion_headers)
    return resp.status_code == 200


def sync_covers_to_book():
    pages_with_empty_cover = _get_pages_with_empty_cover(database_id)
    print(f'当前有{len(pages_with_empty_cover)}本书没有设置封面')
    for page in pages_with_empty_cover:
        book_name = page["book_name"]
        cover = _get_image_from_douban(page["book_name"])
        if cover:
            print(f'成功获取到{book_name}的封面')
            if _set_cover(cover, page["page_id"]):
                print(f'成功设置{book_name}封面')
            else:
                print(f'设置{book_name}封面失败')
        else:
            print(f'未获取到{book_name}的封面')


if __name__ == '__main__':
    sync_covers_to_book()

```

# 前言

我一直有使用 notion 记录自己的[读书笔记](https://leetao-personal.notion.site/leetao-personal/32718a67dc634ab1960b3591bc52d913)的习惯

<img width="909" alt="读书笔记" src="https://user-images.githubusercontent.com/10162120/218299556-c15f2b5b-72be-4d7e-8bef-95df3bd5ee21.png">

在此之前，对应书籍的读书封面，都是我从豆瓣手动复制封面地址，然后在 notion 上设置的，这个流程太过重复和单调了，所以就考虑写个程序将这个流程自动化起来。

# 设置读书模版

如果需要直接使用上述脚本的话，可以直接将我的模版 Duplicate，否则的话，保留封面那一列也可以

<img width="1104" alt="image" src="https://user-images.githubusercontent.com/10162120/218299801-69fc5660-e257-49f0-911d-0da2ff6de210.png">

⚠️封面那一列的属性为 **File & media**
<img width="282" alt="image" src="https://user-images.githubusercontent.com/10162120/218299837-e2e57126-7941-4057-a896-dde1440981ef.png">

# 创建 integrations

去 notion 创建自己的 [integrations](https://www.notion.so/my-integrations)，选择对应的 workspace

<img width="926" alt="截屏2023-02-12 16 07 16" src="https://user-images.githubusercontent.com/10162120/218299946-239dfb83-6720-4f82-8b72-ece265ad4b09.png">

其中 **Capabilities** 中的 **Content Capabilities** 三个都要勾选

<img width="631" alt="image" src="https://user-images.githubusercontent.com/10162120/218299985-c503c2a5-3576-4d24-bfab-3eb6e3950c33.png">

创建完成后，会获得一个 **token**，

<img width="933" alt="image" src="https://user-images.githubusercontent.com/10162120/218300190-814a5c37-c9a5-4ffd-a425-59d8e1cddc3a.png">

# Connect 你的 integrations

在你的图书馆 Connect 你刚才的 integrations

<img width="275" alt="image" src="https://user-images.githubusercontent.com/10162120/218300084-6f274f00-3dc1-48fe-aebe-f6bdb402eaec.png">

**⚠️可能需要过几分钟，才能找到你创建创建的 integrations**

# 设置程序中对应的参数

程序中的 database_id 可以在你的图书馆的 url 找到,这里以我的图书馆为例子，

```
https://leetao-personal.notion.site/leetao-personal/32718a67dc634ab1960b3591bc52d913
```

url 最后一串 uuid 就是 dababase_id，程序中的另外一个参数 token 就是刚才获取的那个 token，notion_version 暂时固定不需要调整。

# 最后

最后将这个脚本放到 github action 上，定时执行一下，就可以不需要人工干预，自动同步封面了；当然程序远不止可以同步封面这些，还可以同步豆瓣的评分，把你之前的阅读记录全部同步到 notion，剩下的这些就可以自己定制化开发了。

**Comment by daheitian** [comment]
挺不错的，小白不会操作，也想同步很多数据 。现在是在用一个插件在同步豆瓣但是就是只能电脑上点。

**Reply by LeetaoGoooo** [reply]
我来抽空做个桌面软件，把整个流程做成自动化，只需要申请 key 和绑定 integrations

**Reply by daheitian** [reply]
太棒了，对于喜欢收集数据的人太友好了

**Comment by Ivan-Mingjr** [comment]
大佬请问下我用这段代码跑了一下，执行结果提示【0本书没有设置封面】，我看字段都搞对了··
代码小白搞不懂，可以请教下吗？
我是把你的代码扒拉下来后，改了databaseid、secret、还有封面图的字段名，直接在pycharm跑的··

**Reply by LeetaoGoooo** [reply]
不介意的话，可以把你的 link 分享一下嘛？然后 code 截图给我，我帮你看一下

**Reply by Ivan-Mingjr** [reply]
我的link：https://www.notion.so/ivanming/Media-5c5c8edceca4488592f902f0ec49cbf1?pvs=4
代码截图：
https://imgse.com/i/piWFa8S
https://imgse.com/i/piWFdgg

**Reply by LeetaoGoooo** [reply]
你试着把 database_id 改成 `7abbc022736e4d38b5596d98bcab038d` 试试？

**Reply by Ivan-Mingjr** [reply]
还是一样的报错···当前有0本书没有设置封面

**Reply by LeetaoGoooo** [reply]
```shell
curl --location 'https://api.notion.com/v1/databases/{{database_id}}/query' \
--header 'Authorization:{{token}}' \
--header 'Content-Type: application/json' \
--header 'Notion-Version: 2022-02-22' \
--data '{
        "filter": {
            "property": "pics",
            "files": {
                "is_empty": true
            }
        }
    }'
```
替换上面的 `database_id` 和 `token`，然后请求看看返回信息是什么，目前看应该就是 `database_id` 找的不对

**Reply by Ivan-Mingjr** [reply]
有点怪异。返回：{"object":"error","status":401,"code":"unauthorized","message":"API token is invalid.","request_id":"a69d3102-ef5e-4fbd-a511-0d9760b2e210"}

核对了好几次secrect肯定是对的，databaseid也试过几个疑似的值，都是返回这个； 最大可能是 7abbc022736e4d38b5596d98bcab038d，可是还是报错

**Reply by LeetaoGoooo** [reply]
这个错误是你的 token 不对呀～ 你的 token 是从 integrations 获取到的嘛？<img width="425" alt="image" src="https://github.com/LeetaoGoooo/discussion-blog/assets/10162120/2b42c7fc-c6d9-4bc8-ba35-86df4bee786b">

**Reply by Ivan-Mingjr** [reply]
我终于解决了--- token是对的，
之所以一直取不到数据，是因为我没把这个我的page添加integrations的connections权限···
notion界面里提示说这个integreations会“works with the XXX workspace”---所以我以为是默认这个space 下的所有page都会自动开启授权···原来是要一个个page打开添加··

现在已经取到了大部分封面图了～～～感谢大佬的时间和热情解答！

**Reply by LeetaoGoooo** [reply]
哈哈 没事，好在问题总算解决了 赞～👍
